syntax = "proto2";

package abi_dump;

message FunctionDecl {
  // Fully Qualified Name.
  required string function_name = 1 [default = "NONE"];

  // Mangled name.
  required string mangled_function_name = 2 [default = "NONE"];

  required string source_file = 3;
  required string parent_name = 4 [default = "NONE"];
  repeated string template_arguments = 5 ;
  repeated FieldDecl parameters = 6;
  required string return_type = 7 [default = "VOID"];
  required string access = 8 [default = "public"];
  required uint32 template_kind = 9 [default = 0];
  required TemplateInfo template_info = 10;
}

message FieldDecl {
  required string field_name = 1 [default = "NONE"];
  required string field_type = 2 [default = "VOID"];
  required string access = 3 [default = "public"];
  required bool default_arg = 4 [default = false];
}

message TemplateInfo {
  repeated FieldDecl template_parameters = 1;
}

message CXXBaseSpecifier {
  required string fully_qualified_name = 1 [default = "NONE"];
  required string access = 2 [default = "public"];
  required bool is_virtual = 3 [default = false];
}

message RecordDecl {
  repeated FieldDecl fields = 2;
  repeated string inner_classes = 3;
  repeated CXXBaseSpecifier base_specifiers = 4;
  required string fully_qualified_name = 5 [default = "NONE"];
  required int64 id = 6 [default = 0];
  required string entity_type = 7 [default = "NONE"];
  required string source_file = 9 [default = "NONE"];
  required bool is_c_struct = 10 [default = false];
  required string template_kind = 11 [default = "NONE"];
  repeated bool template_arguments = 12; //denotes template arguments.
  required TemplateInfo template_info = 13;
}

message TranslationUnit {
  repeated RecordDecl classes = 1;
  repeated FunctionDecl functions = 2;
}
